opm/material/fluidmatrixinteractions/PiecewiseLinearTwoPhaseMaterialParams.hpp

```diff
@@ -33,74 +33,105 @@
 
 #include <opm/material/common/EnsureFinalized.hpp>
 
-namespace Opm {
+namespace Opm
+{
 /*!
  * \ingroup FluidMatrixInteractions
  *
  * \brief Specification of the material parameters for a two-phase material law which
  *        uses a table and piecewise constant interpolation.
  */
-template<class TraitsT>
+template <class TraitsT, class VectorT = std::vector<typename TraitsT::Scalar>>
 class PiecewiseLinearTwoPhaseMaterialParams : public EnsureFinalized
 {
     using Scalar = typename TraitsT::Scalar;
 
 public:
-    using ValueVector = std::vector<Scalar>;
+    using ValueVector = VectorT;
 
     using Traits = TraitsT;
 
     PiecewiseLinearTwoPhaseMaterialParams()
     {
     }
 
+    PiecewiseLinearTwoPhaseMaterialParams(ValueVector SwPcwnSamples,
+                                          ValueVector pcwnSamples,
+                                          ValueVector SwKrwSamples,
+                                          ValueVector krwSamples,
+                                          ValueVector SwKrnSamples,
+                                          ValueVector krnSamples)
+        : SwPcwnSamples_(SwPcwnSamples)
+        , SwKrwSamples_(SwKrwSamples)
+        , SwKrnSamples_(SwKrnSamples)
+        , pcwnSamples_(pcwnSamples)
+        , krwSamples_(krwSamples)
+        , krnSamples_(krnSamples)
+    {
+        finalize();
+    }
+
     /*!
      * \brief Calculate all dependent quantities once the independent
      *        quantities of the parameter object have been set.
      */
     void finalize()
     {
-        EnsureFinalized :: finalize ();
+        EnsureFinalized ::finalize();
 
         // revert the order of the sampling points if they were given
         // in reverse direction.
-        if (SwPcwnSamples_.front() > SwPcwnSamples_.back())
-            swapOrder_(SwPcwnSamples_, pcwnSamples_);
-
-        if (SwKrwSamples_.front() > SwKrwSamples_.back())
-            swapOrder_(SwKrwSamples_, krwSamples_);
+        // Reverting the order involves swapping which only works for non-consts.
+        // The const expr ensures we can create constant parameter views.
+        if constexpr (!std::is_const_v<typename ValueVector::value_type> && !std::is_const_v<ValueVector>) {
```

Comment:
Here we risk breaking the contract of the class that the order of the sw samples is increasing, for example if we use the new constructor to create the object and pass in ranges of invalid data (sw samples decreasing).

I think in such a situation the only correct thing to do is to throw an exception. I propose to move this check to `swapOrder()_`, maybe rename that function to `swapOrderIfInvalidOrThrowIfUnable_()` which is a bit of a mouthful..., and in that function throw if we detect wrong ordering but are unable to deal with it due to immutability.

Reply:
I think this suggestion makes the code more robust, I will encorporate it